sdk/shared/src/main/scala/org/ergoplatform/sdk/ContractTemplate.scala

```diff
@@ -0,0 +1,476 @@
+      w.putOption(data.treeVersion)(_.putUByte(_))
+      serializeString(data.name, w)
+      serializeString(data.description, w)
+
+      val nConstants = data.constTypes.length
+      w.putUInt(nConstants)
+      cfor(0)(_ < nConstants, _ + 1) { i =>
+        TypeSerializer.serialize(data.constTypes(i), w)
+      }
+      w.putOption(data.constValues)((_, values) => {
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          w.putOption(values(i))((_, const) =>
+            DataSerializer.serialize(const, data.constTypes(i), w))
+        }
+      })
+
+      val nParameters = data.parameters.length
+      w.putUInt(nParameters)
+      cfor(0)(_ < nParameters, _ + 1) { i =>
+        Parameter.serializer.serialize(data.parameters(i), w)
+      }
+
+      val expressionTreeWriter = SigmaSerializer.startWriter()
+      ValueSerializer.serialize(data.expressionTree, expressionTreeWriter)
+      val expressionBytes = expressionTreeWriter.toBytes
+      w.putUInt(expressionBytes.length)
+      w.putBytes(expressionBytes)
+    }
+
+    override def parse(r: SigmaByteReader): ContractTemplate = {
+      val treeVersion = r.getOption(r.getUByte().toByte)
+      val name = parseString(r)
+      val description = parseString(r)
+
+      val nConstants = r.getUInt().toInt
+      val constTypes: IndexedSeq[SType] = {
+        if (nConstants > 0) {
+          val res = safeNewArray[SType](nConstants)
+          cfor(0)(_ < nConstants, _ + 1) { i =>
+            res(i) = TypeSerializer.deserialize(r)
+          }
+          res
+        } else {
+          SType.EmptySeq
+        }
+      }
+      val constValues: Option[IndexedSeq[Option[SType#WrappedType]]] = r.getOption((() => {
+        if (nConstants > 0) {
+          val res = safeNewArray[Option[SType#WrappedType]](nConstants)
+          cfor(0)(_ < nConstants, _ + 1) { i =>
+            res(i) = r.getOption((() => DataSerializer.deserialize(constTypes(i), r))())
+          }
+          res
+        } else {
+          Array.empty[Option[SType#WrappedType]]
+        }
+      })())
+
+      val nParameters = r.getUInt().toInt
+      val parameters: IndexedSeq[Parameter] = {
+        if (nParameters > 0) {
+          val res = safeNewArray[Parameter](nParameters)
+          cfor(0)(_ < nParameters, _ + 1) { i =>
+            res(i) = Parameter.serializer.parse(r)
+          }
+          res
+        } else {
+          Parameter.EmptySeq
+        }
+      }
+
+      // Populate constants in constantStore so that the expressionTree can be deserialized.
+      val constants = constTypes.indices.map(i => {
+        val t = constTypes(i)
+        DeserializationSigmaBuilder.mkConstant(defaultOf(t), t)
+      })
+      constants.foreach(c => r.constantStore.put(c)(DeserializationSigmaBuilder))
+
+      val _ = r.getUInt().toInt
+      val expressionTree = ValueSerializer.deserialize(r)
+      CheckDeserializedScriptIsSigmaProp(expressionTree)
```

Comment:
I suggest to not use ValidationRules. They are part of soft-forkability logic and better not to be used outside of that. Use simple type comparison directly here and throw serializer exception.

Reply:
Done.